File: java/src/main/java/org/tvarley/euler/solutions/Solution049.java
/*
 * Solution for Euler problem 049
 *
 * Prime Permutations
 *
 * The arithmetic sequence, 1487, 4817, 8147, in which each of the terms increases by 3330, is unusual in two ways: (i) each of the three terms are prime, and, (ii) each of the 4-digit numbers are permutations of one another.
 * There are no arithmetic sequences made up of three 1-, 2-, or 3-digit primes, exhibiting this property, but there is one other 4-digit increasing sequence.
 * What 12-digit number do you form by concatenating the three terms in this sequence?
 *
 * @author tvarley
 * @date 3/14/2026
 */
package org.tvarley.euler.solutions;

import org.tvarley.euler.Solution;
import org.tvarley.euler.util.Prime;
import java.util.*;

public class Solution049 implements Solution {
  public String solve() {
    for (int a = 1000; a < 10000; a++) {
      if (!Prime.isPrime(a)) continue;
      String sa = String.valueOf(a);
      char[] chars = sa.toCharArray();
      Arrays.sort(chars);
      String sorted = new String(chars);
      List<Integer> perms = new ArrayList<>();
      for (int p : generatePrimes(sorted)) {
        if (String.valueOf(p).length() == 4) perms.add(p);
      }
      Collections.sort(perms);
      for (int i = 0; i < perms.size(); i++) {
        for (int j = i + 1; j < perms.size(); j++) {
          int diff = perms.get(j) - perms.get(i);
          int third = perms.get(j) + diff;
          if (perms.contains(third) && third != perms.get(i) && perms.get(i) != 1487) {
            return "" + perms.get(i) + perms.get(j) + third;
          }
        }
      }
    }
    return "";
  }

  private List<Integer> generatePrimes(String sorted) {
    Set<Integer> primes = new HashSet<>();
    permute(sorted.toCharArray(), 0, primes);
    return new ArrayList<>(primes);
  }

  private void permute(char[] arr, int index, Set<Integer> result) {
    if (index == arr.length) {
      int num = Integer.parseInt(new String(arr));
      if (Prime.isPrime(num)) result.add(num);
      return;
    }
    Set<Character> used = new HashSet<>();
    for (int i = index; i < arr.length; i++) {
      if (used.add(arr[i])) {
        swap(arr, index, i);
        permute(arr, index + 1, result);
        swap(arr, index, i);
      }
    }
  }

  private void swap(char[] arr, int i, int j) {
    char temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }
}

File: javascript/src/euler/solution049.js
// https://projecteuler.net/problem=49

// The arithmetic sequence, 1487, 4817, 8147, in which each of the terms increases by 3330, is unusual in two ways: (i) each of the three terms are prime, and, (ii) each of the 4-digit numbers are permutations of one another.
// There are no arithmetic sequences made up of three 1-, 2-, or 3-digit primes, exhibiting this property, but there is one other 4-digit increasing sequence.
// What 12-digit number do you form by concatenating the three terms in this sequence?

// Answer: 296962999629

const isPrime = (n) => {
  if (n <= 1) return false;
  if (n <= 3) return true;
  if (n % 2 === 0 || n % 3 === 0) return false;
  for (let i = 5; i * i <= n; i += 6) {
    if (n % i === 0 || n % (i + 2) === 0) return false;
  }
  return true;
};

const getPermutations = (arr) => {
  if (arr.length === 0) return [[]];
  const result = [];
  for (let i = 0; i < arr.length; i++) {
    const rest = arr.slice(0, i).concat(arr.slice(i + 1));
    const perms = getPermutations(rest);
    for (const perm of perms) {
      result.push([arr[i], ...perm]);
    }
  }
  return result;
};

module.exports = {
  answer: () => {
    const primes = [];
    for (let i = 1000; i < 10000; i++) {
      if (isPrime(i)) primes.push(i);
    }
    const groups = {};
    for (const p of primes) {
      const key = p.toString().split('').sort().join('');
      if (!groups[key]) groups[key] = [];
      groups[key].push(p);
    }
    for (const group of Object.values(groups)) {
      if (group.length >= 3) {
        group.sort((a, b) => a - b);
        for (let i = 0; i < group.length - 2; i++) {
          for (let j = i + 1; j < group.length - 1; j++) {
            for (let k = j + 1; k < group.length; k++) {
              if (group[j] - group[i] === group[k] - group[j] && group[i] !== 1487) {
                return parseInt(`${group[i]}${group[j]}${group[k]}`);
              }
            }
          }
        }
      }
    }
  }
};